### DISCUSSION TITLE
AuxKernels and ADMaterialPropertiy

### DISCUSSION THREAD

**Discussion by joe61vette** [idaholab/moose#17283] [Q&A General] [answered]
Hi:  
Trying to code an AuxKernel for a Finite Volume based app that consumes material properties that are "AD".  As I understand it, AuxVars are not supported by the AD system.  So, there is no point in doing computations with AD properties inside an AuxKernel.  

The return type for the AuxKernel is simply "Real".  But I want to use an AD Property inside it.  For example, the initialization has the line:

  _rho(getADMaterialProperty<Real>("density")

with the above, I get a compile time error because it can't convert the return type to a double.  If I try to initialize with:

 _rho(getMaterialProperty<Real>("density")

I get a run time error of symbol not found.  

So, how can I use an ADMaterialProperty inside an AuxKernel?

Thanks,
  Joe Kelly

**Comment by ajacquey** [comment, marked as answer]
Hi Joe,

You should add the following include to your AuxKernel src file:
`#include "metaphysicl/raw_type.h"`

and you can access the raw value of your AD material property via:
`MetaPhysicL::raw_value(_rho[_qp])`

I suppose you could also use directly `_rho[_qp].value()` but to be confirmed.

Antoine

**Reply by joe61vette** [reply]
Thanks Antoine.  The _rho[_qp].value() works great.

**Comment by srinath-chakravarthy** [comment]
You are missing a template parameter for the property 
_rho(getADMaterialProperty<Real>("density")) should work if it is declared as ADReal 

Same template is required for the non-AD property as well

**Comment by srinath-chakravarthy** [comment]
oops _rho(getADMaterialProperty<Real>("density"))

**Comment by srinath-chakravarthy** [comment]
_rho(getADMaterialProperty\<Real\>("density"))

**Comment by joe61vette** [comment]
Thanks everyone.  Making sure that I had the template right and using _rho[_qp].value() was the answer.
